### PR TITLE
Implement the --controllers flag fully for the cloud-controller manager

### PR DESCRIPTION
**Pull request by stewart-yu**
**What this PR does / why we need it**:
we had add `--controller` flag for `cloud-controller-manager`, but not implement it. In this PR, we want to implement it, to indicate which controllers are enabled and disabled. From the help:
```
--controllers strings                                                                                             
                A list of controllers to enable. '*' enables all on-by-default controllers, 'foo' enables the controller
                named 'foo', '-foo' disables the controller named 'foo'.
                All controllers: cloudnode, persistentvolume-binder, route, service
                Disabled-by-default controllers:  (default [*])
``` 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by stewart-yu** [comment]
/assign @luxas @sttts 
add `--controller` flag for CCM

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 1]
```diff
@@ -19,31 +19,28 @@ package app
...
 	"k8s.io/apimachinery/pkg/util/uuid"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/apiserver/pkg/server"
 	"k8s.io/apiserver/pkg/server/healthz"
 	apiserverflag "k8s.io/apiserver/pkg/util/flag"
 	"k8s.io/apiserver/pkg/util/globalflag"
-	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/leaderelection"
 	"k8s.io/client-go/tools/leaderelection/resourcelock"
-	cloudprovider "k8s.io/cloud-provider"
```
Comment (luxas) [inline cmd/cloud-controller-manager/app/controllermanager.go]: I'd keep this so one knows what it's imported as.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 2]
```diff
@@ -284,3 +287,28 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+	ret := sets.StringKeySet(NewControllerInitializers())
+	return ret.List()
+}
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+// paired to their InitFunc.  This allows for structured downstream composition and subdivision.
+func NewControllerInitializers() map[string]InitFunc {
```
Comment (luxas) [inline cmd/cloud-controller-manager/app/controllermanager.go]: I guess we have the same pattern in the core crtlmgr? This is good :+1:

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 3]
```diff
@@ -34,18 +32,15 @@ import (
...
+	"k8s.io/cloud-provider"
 	"k8s.io/klog"
 	cloudcontrollerconfig "k8s.io/kubernetes/cmd/cloud-controller-manager/app/config"
 	"k8s.io/kubernetes/cmd/cloud-controller-manager/app/options"
 	genericcontrollermanager "k8s.io/kubernetes/cmd/controller-manager/app"
 	cmoptions "k8s.io/kubernetes/cmd/controller-manager/app/options"
-	cloudcontrollers "k8s.io/kubernetes/pkg/controller/cloud"
-	routecontroller "k8s.io/kubernetes/pkg/controller/route"
-	servicecontroller "k8s.io/kubernetes/pkg/controller/service"
+	kubectrlmgrapp "k8s.io/kubernetes/cmd/kube-controller-manager/app"
```
Comment (luxas) [inline cmd/cloud-controller-manager/app/controllermanager.go]: this is unexpected, did you mean to reference cmd/ccm/app instead?

**Review by luxas — COMMENTED** [review]
Overall, great work, thank you @stewart-yu!

/assign @andrewsykim 
for review as well

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 1 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: you are right, fixed

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 3 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: do a copy in `cmd/ccm/app`, like ` cmd/kcm/app` now

**Comment by andrewsykim** [comment]
Thanks @stewart-yu! We were meaning to do this and this PR looks good overall. At KubeCON this week but will try to give a more extensive review in the next few days.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 4]
```diff
@@ -284,3 +247,54 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+}
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+// paired to their InitFunc.  This allows for structured downstream composition and subdivision.
+func NewControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["cloudnode"] = startCloudNodeController
```
Comment (andrewsykim) [inline cmd/cloud-controller-manager/app/controllermanager.go]: depending on when https://github.com/kubernetes/kubernetes/pull/70344 merges, we may need to add a new `cloud-node-lifecycle` controller here.

Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/controllermanager.go]: also prefer `cloud-node` here.

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed

**Comment by andrewsykim** [comment]
cc @cheftako @mcrute

File: cmd/cloud-controller-manager/app/core.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+				routes,
+				ctx.ClientBuilder.ClientOrDie("route-controller"),
+				ctx.SharedInformers.Core().V1().Nodes(),
+				ctx.ComponentConfig.KubeCloudShared.ClusterName,
+				clusterCIDR,
+			)
+			go routeController.Run(stop, ctx.ComponentConfig.KubeCloudShared.RouteReconciliationPeriod.Duration)
+		}
+	} else {
+		klog.Infof("Will not configure cloud provider routes for allocate-node-cidrs: %v, configure-cloud-routes: %v.", ctx.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs, ctx.ComponentConfig.KubeCloudShared.ConfigureCloudRoutes)
```
Comment (andrewsykim) [inline cmd/cloud-controller-manager/app/core.go]: Should return `nil, false, nil` for this case here?

File: cmd/cloud-controller-manager/app/core.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+	return nil, true, nil
+}
+
+func startRouteController(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (http.Handler, bool, error) {
+
+	// If CIDRs should be allocated for pods and set on the CloudProvider, then start the route controller
+	if ctx.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs && ctx.ComponentConfig.KubeCloudShared.ConfigureCloudRoutes {
+		if routes, ok := cloud.Routes(); !ok {
+			klog.Warning("configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.")
```
Comment (andrewsykim) [inline cmd/cloud-controller-manager/app/core.go]: same here, should we return `nil, false, nil` here?

File: cmd/cloud-controller-manager/app/core.go (hunk 5 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/core.go]: IMO, no need. In this case, we just hint user in the logm but will still start the `RouteController`.

File: cmd/cloud-controller-manager/app/core.go (hunk 6 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/core.go]: the same

File: cmd/cloud-controller-manager/app/core.go (hunk 5 above)
Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/core.go]: I don't follow and maybe I missed something. Here we never actually run the route controller go routine but we will log "Started: routecontroller", that does not seem like the right behavior to me 🤔

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 7]
```diff
@@ -284,3 +247,54 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+// paired to their InitFunc.  This allows for structured downstream composition and subdivision.
+func NewControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["cloud-node"] = startCloudNodeController
+	controllers["persistentvolume-binder"] = startPersistentVolumeLabelController
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/controllermanager.go]: It would be nice if we could find a way to enforce that the controller name is the same in both the KCM and the CCM. I realize you have done this in the code but referencing the same constant or similar would help to ensure that the names are consistent across executables.

**Comment by andrewsykim** [comment]
@stewart-yu https://github.com/kubernetes/kubernetes/pull/70344 was merged, thanks for your patience!

File: cmd/cloud-controller-manager/app/core.go (hunk 5 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/core.go]: you are right. fixed now :+1:

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 8]
```diff
@@ -295,3 +247,55 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+// paired to their InitFunc.  This allows for structured downstream composition and subdivision.
+func NewControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["cloud-node"] = startCloudNodeController
+	controllers["cloud-nodelifecycle"] = startcloudNodeLifecycleController
```
Comment (andrewsykim) [inline cmd/cloud-controller-manager/app/controllermanager.go]: I much prefer this to be `cloud-node-lifecycle` but to be consistent with KCM (and other controllers) it should be `cloudnodelifecycle`. Similar to `cloud-node`, I think I was wrong and maybe we should keep it as `cloudnode` for consistency with other controller names (sorry for the back and forth 😆) .

Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fwiw, `cloudnodelifecycle` is a new controller so might not be too late to change this in the KCM now, will leave it up to you :)

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: notice in KCM before, but for min change, ignore it
fixed now

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 9]
```diff
@@ -387,7 +387,7 @@ func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc
 	if loopMode == IncludeCloudLoops {
 		controllers["service"] = startServiceController
 		controllers["route"] = startRouteController
-		controllers["cloudnodelifecycle"] = startCloudNodeLifecycleController
+		controllers["cloud-node-lifecycle"] = startCloudNodeLifecycleController
```
Comment (andrewsykim) [inline cmd/kube-controller-manager/app/controllermanager.go]: For KCM reviewer: this changed was merged in https://github.com/kubernetes/kubernetes/pull/70344 (i.e. in v1.14) so it is not breaking compatibility for anyone

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 7 above)
Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/controllermanager.go]: I can have a follow up PR for this one :)

**Comment by sttts** [comment]
The flag description is confusing because there is nothing disabled.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 10]
```diff
@@ -216,80 +210,35 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
...
-			time.Sleep(wait.Jitter(c.ComponentConfig.Generic.ControllerStartInterval.Duration, ControllerStartJitter))
-		}
-	} else {
-		klog.Infof("Will not configure cloud provider routes for allocate-node-cidrs: %v, configure-cloud-routes: %v.", c.ComponentConfig.KubeCloudShared.AllocateNodeCIDRs, c.ComponentConfig.KubeCloudShared.ConfigureCloudRoutes)
-	}
-
 	// If apiserver is not running we should wait for some time and fail only then. This is particularly
 	// important when we start apiserver and controller manager at the same time.
-	err = genericcontrollermanager.WaitForAPIServer(c.VersionedClient, 10*time.Second)
+	err := genericcontrollermanager.WaitForAPIServer(c.VersionedClient, 10*time.Second)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: `if err := ...; err != nil {`

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 11]
```diff
@@ -299,6 +248,32 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
 	select {}
 }
 
+// IsControllerEnabled check if a specified controller enabled or not.
+func IsControllerEnabled(name string, disabledByDefaultControllers sets.String, controllers ...string) bool {
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: isn't the flag implementation generic? Why is this logic here and not where the flag is defined?

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 12]
```diff
@@ -299,6 +248,32 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+		if ctrl == "*" {
+			hasStar = true
+		}
+	}
+	// if we get here, there was no explicit choice
+	if !hasStar {
+		// nothing on by default
+		return false
+	}
+	if disabledByDefaultControllers.Has(name) {
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: return !disabledByDefaultControllers.Has(name)

File: cmd/cloud-controller-manager/app/core.go [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog"
+	cloudcontrollerconfig "k8s.io/kubernetes/cmd/cloud-controller-manager/app/config"
+	cloudcontrollers "k8s.io/kubernetes/pkg/controller/cloud"
+	routecontroller "k8s.io/kubernetes/pkg/controller/route"
+	servicecontroller "k8s.io/kubernetes/pkg/controller/service"
+)
+
+func startCloudNodeController(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (http.Handler, bool, error) {
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/core.go]: please use a `Ch` suffix for channels. We usually do that elsewhere, i.e. `stopCh`

File: cmd/cloud-controller-manager/app/core.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+func startcloudNodeLifecycleController(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (http.Handler, bool, error) {
+	// Start the cloudNodeLifecycleController
+	cloudNodeLifecycleController, err := cloudcontrollers.NewCloudNodeLifecycleController(
+		ctx.SharedInformers.Core().V1().Nodes(),
+		ctx.ClientBuilder.ClientOrDie("cloud-node-lifecycle-controller"),
+		cloud,
+		ctx.ComponentConfig.KubeCloudShared.NodeMonitorPeriod.Duration,
+	)
+	if err != nil {
+		klog.Errorf("failed to start cloud node lifecycle controller: %s", err)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/core.go]: looks more like a warning if we continue

File: cmd/cloud-controller-manager/app/options/options.go [hunk 15]
```diff
@@ -253,8 +219,47 @@ func (o *CloudControllerManagerOptions) Config(allControllers, disabledByDefault
 		return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
 	}
 
-	c := &cloudcontrollerconfig.Config{}
-	if err := o.ApplyTo(c, CloudControllerManagerUserAgent); err != nil {
+	kubeconfig, err := clientcmd.BuildConfigFromFlags(o.Master, o.Kubeconfig)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: why was this block of code moved. IMO with the split of logic into Config and ApplyTo it's more complex now.

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 12 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: seems no need:  if `disabledByDefaultControllers.Has(name)` is `true`, will return `false`, else will return `true` in line276

File: cmd/cloud-controller-manager/app/options/options.go (hunk 15 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: revert now

File: cmd/cloud-controller-manager/app/core.go (hunk 14 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/core.go]: fixed

File: cmd/cloud-controller-manager/app/core.go (hunk 13 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/core.go]: fixed

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 11 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: Follow KCM. may be we can move it to `cmd/controller-manger` , so we can reuse it both in KCM and CCM

**Comment by liggitt** [comment]
linking PR's that had a test flake because of https://github.com/kubernetes/kubernetes/issues/71696

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 11 above)
Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/controllermanager.go]: That would be ideal but a little more refactoring will be required. Might be cleaner to have that in another PR, just my 2 cents :)

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed, move `IsControllerEnabled()` into `cmd/controller-manager/app/helper.go`

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 16]
```diff
@@ -213,85 +212,71 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
...
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// newControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+// paired to their InitFunc.  This allows for structured downstream composition and subdivision.
+func newControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["cloud-node"] = startCloudNodeController
+	controllers["cloud-node-lifecycle"] = startcloudNodeLifecycleController
```
[the remark this answers is not in the document]
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed :+1:

File: cmd/controller-manager/app/helper.go [hunk 17]
```diff
@@ -53,3 +54,26 @@ func WaitForAPIServer(client clientset.Interface, timeout time.Duration) error {
 
 	return nil
 }
+
+// IsControllerEnabled check if a specified controller enabled or not.
+func IsControllerEnabled(name string, disabledByDefaultControllers sets.String, controllers ...string) bool {
```
Comment (sttts) [inline cmd/controller-manager/app/helper.go]: the signature is slightly odd: why a ... variable for controllers? It's not a variadic usage here.

Better:
```
func IsControllerEnabled(name string, all, disabledByDefault sets.String) bool {
```

File: cmd/controller-manager/app/helper_test.go [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2018 The Kubernetes Authors.
```
Comment (sttts) [inline cmd/controller-manager/app/helper_test.go]: leave the date

File: cmd/controller-manager/app/helper_test.go [hunk 19]
```diff
@@ -77,5 +73,4 @@ func TestIsControllerEnabled(t *testing.T) {
 		actual := IsControllerEnabled(tc.controllerName, sets.NewString(tc.disabledByDefaultControllers...), tc.controllers...)
 		assert.Equal(t, tc.expected, actual, "%v: expected %v, got %v", tc.name, tc.expected, actual)
 	}
-
```
Comment (sttts) [inline cmd/controller-manager/app/helper_test.go]: remove changes in this file

File: cmd/controller-manager/app/helper.go (hunk 17 above)
Comment (sttts) [reply cmd/controller-manager/app/helper.go]: or `func IsControllerEnabled(name string, all []string, disabledByDefault sets.String) bool {` if that fits better.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 20]
```diff
@@ -215,86 +212,45 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
...
+		if !started {
+			klog.Warningf("Skipping %q", controllerName)
+			continue
+		}
+		klog.Infof("Started %q", controllerName)
 
-			routeController := routecontroller.New(routes, client("route-controller"), c.SharedInformers.Core().V1().Nodes(), c.ComponentConfig.KubeCloudShared.ClusterName, clusterCIDR)
-			go routeController.Run(stop, c.ComponentConfig.KubeCloudShared.RouteReconciliationPeriod.Duration)
-			time.Sleep(wait.Jitter(c.ComponentConfig.Generic.ControllerStartInterval.Duration, ControllerStartJitter))
+		if debugHandler != nil && unsecuredMux != nil {
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: where does this change come from inside of a refactoring commit?

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 21]
```diff
@@ -148,15 +143,17 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
 	}
 
 	// Start the controller manager HTTP server
+	// unsecuredMux is the handler for these controller *after* authn/authz filters have been applied
+	var unsecuredMux *mux.PathRecorderMux
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: this change seems to be unrelated to the refactoring. But even it belongs here, I doubt this is correct: this way the controller debug paths are installed only into the insecure serving mux if it is enabled.

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: sorry, my mistakes. The debug function no need for all controller in CCM, just see in `GCcontroller` in KCM.
will revert at once

File: cmd/controller-manager/app/helper.go (hunk 17 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/helper.go]: accept

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 21 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 22]
```diff
@@ -295,3 +298,29 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+// KnownControllers indicate the default controller we are known.
+func KnownControllers() []string {
+	ret := sets.StringKeySet(newControllerInitializers())
+	return ret.List()
+}
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// newControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: not so public anymore, comment seems to be outdated.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 23]
```diff
@@ -213,7 +215,7 @@ func Run(c *cloudcontrollerconfig.CompletedConfig, stopCh <-chan struct{}) error
 }
 
 // startControllers starts the cloud specific controller loops.
-func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan struct{}, cloud cloudprovider.Interface) error {
+func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan struct{}, cloud cloudprovider.Interface, controllers map[string]InitFunc) error {
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: nit: this change belong into the other commit.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 24]
```diff
@@ -295,3 +298,29 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
 
 	select {}
 }
+
+// InitFunc is used to launch a particular controller.  It may run additional "should I activate checks".
+// Any error returned will cause the controller process to `Fatal`
+// The bool indicates whether the controller was enabled.
+type InitFunc func(ctx *cloudcontrollerconfig.CompletedConfig, cloud cloudprovider.Interface, stop <-chan struct{}) (debuggingHandler http.Handler, enabled bool, err error)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: is this still needed outside of this package? Make it private if not.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 25]
```diff
@@ -295,3 +298,29 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, stop <-chan stru
...
+	ret := sets.StringKeySet(newControllerInitializers())
+	return ret.List()
+}
+
+// ControllersDisabledByDefault is the controller disabled default when starting cloud-controller managers.
+var ControllersDisabledByDefault = sets.NewString()
+
+// newControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+// paired to their InitFunc.  This allows for structured downstream composition and subdivision.
+func newControllerInitializers() map[string]InitFunc {
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: does this have to be a func? Wouldn't a private var at the top be enough?

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: seems using function looks better

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 24 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 23 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fixed

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 22 above)
Comment (sttts) [reply cmd/cloud-controller-manager/app/controllermanager.go]: still open

File: cmd/cloud-controller-manager/app/core.go [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+	if !ok {
+		klog.Warning("configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.")
+		return nil, false, nil
+	}
+	var clusterCIDR *net.IPNet
+	var err error
+	if len(strings.TrimSpace(ctx.ComponentConfig.KubeCloudShared.ClusterCIDR)) != 0 {
+		_, clusterCIDR, err = net.ParseCIDR(ctx.ComponentConfig.KubeCloudShared.ClusterCIDR)
+		if err != nil {
+			klog.Warningf("Unsuccessful parsing of cluster CIDR %v: %v", ctx.ComponentConfig.KubeCloudShared.ClusterCIDR, err)
```
Comment (tedyu) [inline cmd/cloud-controller-manager/app/core.go]: Parsing failed.
Yet we still pass clusterCIDR on line 126 below ?

Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/core.go]: follow the original code
besides, If parsing failed, the `clusterCIDR` just `nil`
